trust/builder.c

```diff
@@ -339,6 +339,69 @@ type_der_ext (p11_builder *builder,
 	return check_der_struct (builder, "PKIX1.Extension", attr);
 }
 
+static bool
+type_false_or_time (p11_builder *builder,
+		    CK_ATTRIBUTE *attr)
+{
+	struct tm tm;
+	char *value;
+
+	if (sizeof (CK_BBOOL) == attr->ulValueLen &&
+	    *((CK_BBOOL *)attr->pValue) == CK_FALSE)
+		return true;
+
+	value = attr->pValue;
+
+	switch (attr->ulValueLen) {
+	case 13:
+		/* UTCTime restricted by RFC 5280 4.1.2.5.1, i.e., in
+		 * the format "YYMMDDHHMMSSZ" */
+		if (value[attr->ulValueLen - 1] != 'Z')
+			return false;
+
+		tm.tm_year = atoin (value, 2);
+		if (tm.tm_year < 0)
+			return false;
+		if (tm.tm_year >= 50)
+			tm.tm_year += 1900;
+		else if (tm.tm_year >= 0)
+			tm.tm_year += 2000;
+		value += 2;
+
+		break;
+	case 15:
+		/* GeneralizedTime restricted by RFC 5280 4.1.2.5.2,
+		 * i.e., in the form "YYYYMMDDHHMMSSZ" */
+		if (value[attr->ulValueLen - 1] != 'Z')
+			return false;
+
+		tm.tm_year = atoin (value, 4);
+		if (tm.tm_year < 0)
+			return false;
+		value += 4;
+
+		break;
+	default:
+		return false;
+	}
+
+	tm.tm_mon = atoin (value, 2);
+	value += 2;
+	tm.tm_mday = atoin (value, 2);
+	value += 2;
+	tm.tm_hour = atoin (value, 2);
+	value += 2;
+	tm.tm_min = atoin (value, 2);
+	value += 2;
+	tm.tm_sec = atoin (value, 2);
+
+	if (tm.tm_mon <= 0 || tm.tm_mday <= 0 ||
+	    tm.tm_hour < 0 || tm.tm_min < 0 || tm.tm_sec < 0)
+		return false;
```

Comment:
What about the other extremes? E.g., `tm_hour >= 24`

Reply:
I was supposed to add a check using `mktime` below, but completely forgot about it. Good catch, fixed.